EmbedSOM+optimizations

exaexa opened the pull request:
EmbedSOM+optimizationsSummary (commit messages):

```
commit 082a9eff5bd42070448d38ae67aa428e21ec3124

    High-level performance optimization
    
    Mainly including:
    - use of NearestNeighbors trees for finding the BMU
    - abuse linearity to change the neighborhood update calculation from
      expand-then-sum to more efficient sum-then-expand

commit b280b87f7d61f5cbc71bb703d4c86fbc778fd02b

    Implement EmbedSOM dimensionality reduction
```

Note: I rebased my changes atop the new `getRadius` functionality and it seems compatible, but please check that I merged it correctly and everything works as intended.

Thanks!
-mk

laurentheirendt commented on Project.toml:
```diff
@@ -12,6 +12,7 @@ Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
 FCSFiles = "d76558cf-badf-52d4-a17e-381ab0b0d937"
```
Why might not need `Distributions` anymore.

laurentheirendt commented on src/core.jl:
```diff
@@ -61,66 +61,69 @@ end
+
     if rStart == 0.0
-        rStart = √(som.xdim^2 + som.ydim^2) / 2
+        rStart = √(som.xdim^2 + som.ydim^2) / 3
```
Why `/ 3`?

oHunewald replied:
/ 2 was from the original som implementation, in my view it is a good starting radius if not specified otherwise

laurentheirendt replied:
@oHunewald, so is `/2` more appropriate than `/3`?

oHunewald replied:
I would like to keep /2 as this has been our default radius all the time and gives reasonable results. Otherwise one can pass different start radius as parameter

laurentheirendt commented on src/core.jl:
```diff
@@ -130,46 +133,30 @@ end
-- `kernelFun`: distance kernel function of type fun(x, r)
-- `r`: training radius
-- `toroidal`: if true, the SOM is toroidal.
+- `nnTree`: knn-compatible tree built upon the codes
```
Maybe `tree` is a better variable name.

laurentheirendt commented on src/core.jl:
```diff
@@ -236,15 +223,23 @@ Return a new neighbourhood radius
+function getRadius(initRadius::Float64, finalRadius::Float64, iteration::Int64, decay::String, epochs::Int64)
+
+    # prevent various mathematical trouble
+    @assert initRadius > 0
```
Within functions, it is better to use true if conditions and customize the error message.

laurentheirendt commented on src/core.jl:
```diff
@@ -236,15 +223,23 @@ Return a new neighbourhood radius
+
+    scaledTime = Float64(iteration-1) / Float64(epochs)
+
+    if decay == "linear"
```
What if `decay` is neither `linear` nor `exp`?

laurentheirendt commented on src/satellites.jl:
```diff
@@ -31,16 +31,15 @@ end
-
-    return Distributions.pdf.(Distributions.Normal(0.0,r/3), x)
+function gaussianKernel(x, r::Float64)
+    return exp.(x.^2 .* (-1 / r^2))
```
nice! 👍

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+        end
+
+        #0.000001 avoids hitting zeroes
+        sum = -ssum / (0.00000001 + sum * boost)
```
Using scientific notation might be better here.

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+    end
+end
+
+function embedGigaSOM_internal(som::GigaSOM.Som, data::Array{Float64,2}, t, k, adjust, boost)
```
Is there a function documentation somewhere?

laurentheirendt commented on src/embedding.jl:
````diff
@@ -0,0 +1,172 @@
+e = embedGigaSOM(som, data)
+```
+
+Plotting is best done using histograms; e.g. with Gadfly:
````
Is this description of plotting related to `embedGigaSOM`? Maybe this would fit better in a tutorials and or the visualization folder

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+Data must have the same number of dimensions as the training dataset
+and will be normalised with the same parameters.
+"""
+function embedGigaSOM(som::GigaSOM.Som, data::DataFrame;
```
Can you add a few more comments to this function? Hard to follow ...

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+    if nWorkers > 1
+        dData = distribute(data)
+
+        dRes = [ (@spawnat w embedGigaSOM_internal(som, localpart(dData), t, k, adjust, boost)) for w in workers() ]
```
I think `enumerate(workers)` would be safer.

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+        dRes = [ (@spawnat w embedGigaSOM_internal(som, localpart(dData), t, k, adjust, boost)) for w in workers() ]
+
+        #hopefully the data are separated to localparts in correct order...
+        return vcat([fetch(r) for r in dRes]...)
```
nice! 👍 The order will be ok if the workers have been enumerated (see comment above)

exaexa replied:
Good to know, thanks a lot!

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+            ip = kndist[i]
+
+            # a bit of gravity to avoid singularities
+            gs = 0.00001 * ip
```
Scientific notation

laurentheirendt commented on src/embedding.jl:
```diff
@@ -0,0 +1,172 @@
+            # a bit of gravity to avoid singularities
+            gs = 0.00001 * ip
+
+            mtx[1,1] += gs
```
I would need a few more comments and/or more self-explaining variable names to follow properly here.

exaexa commented:
@laurentheirendt I have hopefully fixed most of the issues you raised, thanks again for review.

Additionally, you commented on the `getRadius` function internals that @oHunewald added into `develop`; which I rewrote later anyway (in PR #55, now closed). For this reason I've moved this rewrite here as well, since there's probably no reason in postponing the fixes. All related commits should be nicely separated from the rest.

Also, since PR #56 is comparatively tiny and just follows the pattern introduced in the other changes here, I merged that as well (that should be the last commit in the series now).

Sorry for expanding the PR, but since all the changes are very related it's probably easier&more meaningful this way.

oHunewald commented on src/satellites.jl:
```diff
@@ -31,16 +31,15 @@ end
 """
-function gaussianKernel(x::Array{Float64, 1}, r::Float64)::Array{Float64, 1}
-
-    return Distributions.pdf.(Distributions.Normal(0.0,r/3), x)
```
by removing Distributions.pdf I loose sensitivity. I don't see a significant performance issue. Run time is almost identical

oHunewald commented on src/core.jl:
```diff
@@ -188,15 +189,17 @@ function mapToGigaSOM(som::Som, data::DataFrame)
          @sync for (p, pid) in enumerate(workers())
              @async R[p] = @spawnat pid begin
-                 visual(som.codes, localpart(dData))
+                 vcat(knn(nnTree, transpose(localpart(dData)), 1)[1]...)
```
just a short comment what is done here

oHunewald commented on src/GigaSOM.jl:
```diff
@@ -11,14 +11,15 @@ module GigaSOM
     using DataFrames
     using Distances
     using Distributed
-    using Distributions
```
I'd like to keep the original kernel function

laurentheirendt commented:
The tests are not passing yet ...

exaexa commented:
Oh, it seems that I neglected one change in the code that is actually breaking the new way of computing gaussians -- the `gaussianKernel` is strictly typed to process 1D float arrays, but is in fact generic even to 2D arrays. Simplest fix is to remove the type annotations (I'll push that ASAP), but I don't know if that is desirable.

Is there any way to say "Any Float array type" in julia, so that the typechecking stays there?

Thanks!

laurentheirendt commented:
@exaexa, just leave the strict typing out for now ...

oHunewald commented:
I have always tested by starting radius

rStart = √(som.xdim^2 + som.ydim^2) / 2

I don't see a reason to change that. Otherwise one can pass a starting
radius as paramter

Am Di., 3. Sept. 2019 um 10:54 Uhr schrieb Laurent Heirendt <
notifications@github.com>:

> *@laurentheirendt* commented on this pull request.
> ------------------------------
>
> In src/core.jl
> <https://github.com/LCSB-BioCore/GigaSOM.jl/pull/54#discussion_r320160861>
> :
>
> >      if rStart == 0.0
>
> -        rStart = √(som.xdim^2 + som.ydim^2) / 2
>
> +        rStart = √(som.xdim^2 + som.ydim^2) / 3
>
>
> @oHunewald <https://github.com/oHunewald>, so is /2 more appropriate than
> /3?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LCSB-BioCore/GigaSOM.jl/pull/54?email_source=notifications&email_token=AANQA2327UYAZ6FOXA2Q5XLQHYQ3XA5CNFSM4ISLGZS2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDN7MGQ#discussion_r320160861>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANQA26DYZ3LFOHMZ7XGR4LQHYQ3XANCNFSM4ISLGZSQ>
> .
>

laurentheirendt commented:
may you change that? thanks.

exaexa commented:
> may you change that? thanks.

It is already removed, reverted it a few commits back when reverting the gaussians.

laurentheirendt commented:
Thanks @exaexa and @oHunewald. Great work! 👍
